diff --git a/Directory.Packages.props b/Directory.Packages.props
index 5efad590..1641603a 100644
--- a/Directory.Packages.props
+++ b/Directory.Packages.props
@@ -7,8 +7,8 @@
   <ItemGroup>
     <PackageVersion Include="Amazon.AspNetCore.DataProtection.SSM" Version="3.2.1" />
     <PackageVersion Include="Aspire.Hosting.AppHost" Version="8.1.0" />
-    <PackageVersion Include="AWSSDK.SecurityToken" Version="3.7.400.10" />
-    <PackageVersion Include="AWSSDK.SimpleSystemsManagement" Version="3.7.401.8" />
+    <PackageVersion Include="AWSSDK.SecurityToken" Version="3.7.400.11" />
+    <PackageVersion Include="AWSSDK.SimpleSystemsManagement" Version="3.7.401.9" />
     <PackageVersion Include="BenchmarkDotNet" Version="0.14.0" />
     <PackageVersion Include="coverlet.msbuild" Version="6.0.2" />
     <PackageVersion Include="GitHubActionsTestLogger" Version="2.4.1" />
